### ISSUE TITLE
mikktspace based tangent generation

### ISSUE THREAD

**Issue by Metric** [assimp/assimp#2363]
It does not appear Assimp uses mikktspace generation for tangents. I highly recommend adding mikktspace generation to align better with the 3D industry. It can be added in as an extra option for those that want to use it, and it does not need to replace the current tangent generation.

copy of mikktspace.c: https://svn.blender.org/svnroot/bf-blender/trunk/blender/intern/mikktspace/mikktspace.c

copy of mikktspace.h: https://svn.blender.org/svnroot/bf-blender/trunk/blender/intern/mikktspace/mikktspace.h

Original paper: http://image.diku.dk/projects/media/morten.mikkelsen.08.pdf

Applications currently using mikktspace: Blender, Unreal Engine 4, Unity, xNormal, Knald Products, Allegorithmic Products, Maya, and Houdini (via plugin). Those are the major ones I know of that are using it or have added it as an option.

**Comment by JekaS178RUS** [comment]
Voting for this! This is very very good idea!

**Comment by JekaS178RUS** [comment]
@kimkulling Hello Kim!
I am thinking of adding mikktspace post process to assimp.
There is a difference between how assimp store tangents and mikktspace suggest to do this is: assimp uses vec3 for tangent and vec3 for bitangent while mikktspace use vec3 for tangent and single float for sign to later calculate bitangent as B = sign * cross(N, T) in fragment shader. As far as I can see usually developers pack tangent and sign in single vec4 vertex attribute. So here I need your opinion how to deal with difference. I think best option is to add new field aiVector4D* mMikkTangents to aiMesh struct.  This will backward compatible with existing code base using assimp, but I am not sure how it will work with importers and exporters.

**Comment by turol** [comment]
That would change the size of the `aiMesh` struct and so would be an ABI change and would required a major version bump.

**Comment by JekaS178RUS** [comment]
@turol good point, what we can do as instead? Put sign to bitangent x component? But how to know if it is actual vector or just sign on export for example?

**Comment by Metric** [comment]
Actually, mikktspace also has a m_SetTSpace in the context that provides the full vec3 bitangent. So nothing really needs to change in the Mesh class of Assimp. If you don't want the full bitangent, then you can also just use the bitangent field of the Mesh to store the sign, rather than the full bitangent.

There can be a post processing option for that as to which context to use when setting the tangent from mikktspace. Aka a post process to include full bitangents in the bitangents field or just store the sign in the bitangents field. I think that would be a good compromise.

**Comment by kimkulling** [comment]
We could think about a configuration flag which requests how to store the data. And depending on this we could use the data in the historical way or in the way mikktspace needs to use them ( vec3 + one float if I get it right ). 
Or we can think about not changing the existing code by such a hack. We can aso think about a more flexible way for data in meshes. Any opinions?

**Comment by JekaS178RUS** [comment]
We can introduce two post process instead of one: 
aiProcess_CalcMikkTangentSpace for T+B
aiProcess_CalcMikkBasicTangentSpace for T+sign
It will match mikktspace.h interface, keep aiMesh unchanged and will let user ability choose which one to use.
Only problem that user will have to repack vec3 + float arrays to a single vec4 array.

From the other side how we should threat this information when tangent space is read/write from/to file?

**Comment by inhosens** [comment]
I think storing bitangent is not necessary. It's very sitty. For example, gltf2 has T + sign, if someone want to read gltf2 through assimp, assimp should generate bitangent from original data. The worse thing happens when users process it. Normally making a bitangent in a vertex shader needs 2 vector multiplications and 1 addition but using the input bitangent needs a matrix transformation. I think it's not reasonable without considering one additional attribute buffer. I vote to T+sign.

**Comment by tellypresence** [comment]
Any updates on this?

**Comment by kimkulling** [comment]
There is a branch.

**Comment by tellypresence** [comment]
Feature request issues tracked via epic #6183
